app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/rate/SwapRateTickerManager.kt

```diff
@@ -0,0 +1,133 @@
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.plus
+import org.p2p.core.utils.Constants.USD_READABLE_SYMBOL
+import org.p2p.core.utils.scaleLong
+import org.p2p.wallet.swap.jupiter.interactor.model.SwapTokenModel
+import org.p2p.wallet.swap.jupiter.statemanager.SwapCoroutineScope
+import org.p2p.wallet.swap.jupiter.statemanager.SwapState
+import org.p2p.wallet.swap.model.jupiter.SwapRateTickerState
+import org.p2p.wallet.user.repository.UserLocalRepository
+import org.p2p.wallet.user.repository.prices.TokenId
+import org.p2p.wallet.user.repository.prices.TokenPricesRemoteRepository
+
+class SwapRateTickerManager private constructor(
+    swapScope: SwapCoroutineScope,
+    private val userLocalRepository: UserLocalRepository,
+    private val tokenPricesRepository: TokenPricesRemoteRepository,
+    private val initDispatcher: CoroutineDispatcher
+) : CoroutineScope by (swapScope + initDispatcher) {
+
+    constructor(
+        swapScope: SwapCoroutineScope,
+        userLocalRepository: UserLocalRepository,
+        tokenPricesRepository: TokenPricesRemoteRepository
+    ) : this(
+        swapScope = swapScope,
+        userLocalRepository = userLocalRepository,
+        tokenPricesRepository = tokenPricesRepository,
+        initDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+    )
+
+    private var currentTokenA: SwapTokenModel? = null
+    private var currentTokenB: SwapTokenModel? = null
+
+    private val currentRateState = MutableStateFlow<SwapRateTickerState>(SwapRateTickerState.Loading)
+
+    fun observe(): Flow<SwapRateTickerState> = currentRateState
+
+    fun handleRoutesLoading(state: SwapState.LoadingRoutes) {
+        currentRateState.value = SwapRateTickerState.Loading
+    }
+
+    fun handleJupiterRates(state: SwapState.SwapLoaded) {
+        val newTokenA = state.tokenA.also { currentTokenA = it }
+        val newTokenB = state.tokenB.also { currentTokenB = it }
+
+        val result = when {
+            !newTokenA.isStable() && newTokenB.isStable() -> {
+                val newRate = (state.amountTokenB / state.amountTokenA).scaleLong(newTokenB.decimals)
+                "1 ${newTokenA.tokenSymbol} ≈ $newRate ${newTokenB.tokenSymbol}"
+            }
+            else -> {
+                val newRate = (state.amountTokenA / state.amountTokenB).scaleLong(newTokenA.decimals)
+                "1 ${newTokenB.tokenSymbol} ≈ $newRate ${newTokenA.tokenSymbol}"
+            }
+        }
+
+        currentRateState.value = SwapRateTickerState.Shown(result)
+    }
+
+    fun onInitialTokensSelected(
+        newTokenA: SwapTokenModel?,
+        newTokenB: SwapTokenModel?
+    ) {
+        if (newTokenA == null || newTokenB == null) return
+
+        val currentTokenSymbolA = currentTokenA?.tokenSymbol
+        val currentTokenSymbolB = currentTokenB?.tokenSymbol
+        if (newTokenA.tokenSymbol == currentTokenSymbolA && newTokenB.tokenSymbol == currentTokenSymbolB) {
```

Comment:
should we ignore CASE for symbols when using equals? 
because SwapTokenModel can be UserToken (Token.Active from solanalist.json) and JupiterToken (from jupiter API). maybe there is a change that they will have different naming. 
maybe even better - check for mint addresses instead of symbols

Reply:
sounds good, will do